include/sys/dmu.h

```diff
@@ -281,9 +281,30 @@ typedef enum dmu_object_type {
  * the transaction is full. See the comment above dmu_tx_assign() for more
  * details on the meaning of these flags.
  */
-#define	DMU_TX_NOWAIT		(0ULL)
-#define	DMU_TX_WAIT		(1ULL<<0)
-#define	DMU_TX_NOTHROTTLE	(1ULL<<1)
+typedef enum {
+	/*
+	 * If the tx cannot be assigned to a transaction for any reason, do
+	 * not block but return immediately.
+	 */
+	DMU_TX_NOWAIT		= 0,
+
+	/*
+	 * Assign the tx to the open transaction. If the open transaction is
+	 * full, or the write throttle is active, block until the next
+	 * transaction and try again. If the pool suspends while waiting,
+	 * return an error.
```

Comment:
It would be good if the comments here made it clear that TX_SUSPEND requires TX_WAIT, and that TX_WAIT's error behavior here is only with TX_SUSPEND not set.

Reply:
Last push tries to improve it a little. It's still a lot to wrap head around.